content/en/blog/_posts/2024-02-29-cri-o-seccomp-oci-artifacts.md

```diff
@@ -0,0 +1,488 @@
+cool feature, the majority of end users would like to link seccomp profiles to
+published container images. This can be done by using a container image annotation; instead
+of being applied to a Kubernetes Pod, the annotation is some metadata applied
+at the container image itself. For example, [Podman](https://podman.io) can be
+used to add the image annotation directly during image build:
+
+```shell
+podman build \
+    --annotation seccomp-profile.kubernetes.cri-o.io=quay.io/crio/seccomp:v1 \
+    -t quay.io/crio/nginx-seccomp:generic .
+```
+
+The pushed image then contains the annotation:
+
+```shell
+skopeo inspect --raw docker://quay.io/crio/nginx-seccomp:generic |
+    jq '.annotations."seccomp-profile.kubernetes.cri-o.io"'
+```
+
+```console
+"quay.io/crio/seccomp:v1"
+```
+
+If I now use that image in an CRI-O test pod definition:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod
+  # no Pod annotations set
+spec:
+  containers:
+    - name: container
+      image: quay.io/crio/nginx-seccomp:generic
+```
+
+Then the CRI-O logs will indicate that the image annotation got evaluated and the
+profile got applied:
+
+```shell
+kubectl delete pod/pod
+```
+
+```console
+pod "pod" deleted
+```
+
+```shell
+kubectl apply -f pod.yaml
+```
+
+```console
+pod/pod created
+```
+
+```console
+INFO[…] Found image specific seccomp profile annotation: seccomp-profile.kubernetes.cri-o.io=quay.io/crio/seccomp:v1  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Pulling OCI artifact from ref: quay.io/crio/seccomp:v1  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Trying to read profile from: /var/lib/containers/storage/overlay/c15f3eb1efd9abe0ae973e050c374988bba52c45a53f5c3cd80419a865013fa6/merged/seccomp.json  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Retrieved OCI artifact seccomp profile of len: 18853  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Created container 116a316cd9a11fe861dd04c43b94f45046d1ff37e2ed05a4e4194fcaab29ee63: default/pod/container  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+```
+
+```shell
+export CONTAINER_ID=$(sudo crictl ps --name container -q)
+sudo crictl inspect $CONTAINER_ID | jq .info.runtimeSpec.linux.seccomp | head
+```
+
+```yaml
+{
+  "defaultAction": "SCMP_ACT_ERRNO",
+  "defaultErrnoRet": 38,
+  "architectures": [
+    "SCMP_ARCH_X86_64",
+    "SCMP_ARCH_X86",
+    "SCMP_ARCH_X32"
+  ],
+  "syscalls": [
+    {
+```
+
+For container images, the annotation `seccomp-profile.kubernetes.cri-o.io` will
+be treated in the same way as `seccomp-profile.kubernetes.cri-o.io/POD` and
+applies to the whole pod. In addition to that, the whole feature also works when
+using the container specific annotation on an image, for example if a container
+is named `container1`:
```

Comment:
There's a risk of confusion: we're talking about two things that are both called container images:
- container images that you can run with kubectl exec
- seccomp profiles that are bundled into container images that you can't run, but marked with metadata to suggest you could

A reader might not know which use you mean here. I genuinely think that assigning a media type to these OCI objects containing seccomp profiles would be the easy path here.

Reply:
That's the plan. Resolving this discussion now.